storage_controller/src/service.rs

```diff
@@ -2744,6 +2745,115 @@ impl Service {
+                    .get(&node_id)
+                    .expect("Pageservers may not be deleted while referenced");
+
+                targets.push((*tenant_shard_id, node.clone()));
+            }
+            targets
+        };
+
+        if targets.is_empty() {
+            return Err(ApiError::NotFound(
+                anyhow::anyhow!("Tenant not found").into(),
+            ));
+        }
+
+        async fn detach_one(
+            tenant_shard_id: TenantShardId,
+            timeline_id: TimelineId,
+            node: Node,
+            jwt: Option<String>,
+        ) -> Result<(ShardNumber, models::detach_ancestor::AncestorDetached), ApiError> {
+            tracing::info!(
+                "Detaching timeline on shard {tenant_shard_id}/{timeline_id}, attached to node {node}",
+            );
+
+            let client = PageserverClient::new(node.get_id(), node.base_url(), jwt.as_deref());
+            client
+                .timeline_detach_ancestor(tenant_shard_id, timeline_id)
+                .await
+                .map_err(|e| {
+                    use mgmt_api::Error;
+
+                    match e {
+                        // no ancestor (ever)
+                        Error::ApiError(StatusCode::CONFLICT, msg) => {
+                            ApiError::Conflict(format!("{node}: {msg}"))
+                        }
+                        // too many ancestors
+                        Error::ApiError(StatusCode::BAD_REQUEST, msg) => {
+                            ApiError::BadRequest(anyhow::anyhow!("{node}: {msg}"))
+                        }
+                        // rest can be mapped
+                        other => passthrough_api_error(&node, other),
+                    }
+                })
+                .map(|res| (tenant_shard_id.shard_number, res))
+        }
+
+        // no shard needs to go first/last; the operation should be idempotent
+        // TODO: it would be great to ensure that all shards return the same error
+        let mut results = self
+            .tenant_for_shards(targets, |tenant_shard_id, node| {
+                futures::FutureExt::boxed(detach_one(
+                    tenant_shard_id,
+                    timeline_id,
+                    node,
+                    self.config.jwt_token.clone(),
+                ))
+            })
+            .await?;
+
+        let any = results.pop().expect("we must have at least one response");
+
+        // FIXME: the ordering is not stable yet on pageserver, should be (ancestor_lsn,
+        // TimelineId)
+        let mismatching = results
+            .iter()
+            .filter(|(_, res)| res != &any.1)
+            .collect::<Vec<_>>();
+        if !mismatching.is_empty() {
+            let matching = mismatching.len() - (results.len() + 1);
+            tracing::error!(
+                matching,
+                compared_against=?any,
+                ?mismatching,
+                "shards returned different results"
+            );
```

Comment:
Informational ... I would be very surprised if this ever happened. This PRs version will randomly produce such results because the order in the reparented_timelines is unspecified, but it will be fixed in the next PR. Alternatively could had treated this as a set... Unsure why I didn't want that, could still change it for the next PR.

Reply:
At the same time, I cannot make this into an assertion either, because I will want to add tests which exercise this path by mixing in an a timeline deletion in between retries of timeline ancestor detach.

Reply:
I was alluding more to the case where the detach succeeds on some shards, but fails on one for some reason. It's fine for this log line to be informational, but I'm curious what one should do in such a scenario. Keep retrying I guess, since all errors should be transient.

Reply:
Hmm you are correct this should probably a warning as there is no clear action, and the warn would give out a test error just as well.

On this line we are also on the happy path so yes actually I am unsure if a test could even be made to exercise this as we currently exit on the first error (pageserver shutdowns and before next retry a timeline is deleted) so we wouldn't get to here.

Another thing I was thinking of that we should return the reparented of which all agree... But I will need to think about this more.

Reply:
There is in fact, a clear action.

In the next PR, I've added a test which hits this `error!`. It can easily happen if you have timeline deletion while timeline detach is ongoing:
- first node will report the to-be-deleted timeline as reparented
- other one will delete the timeline before reparenting

There is no action required here. However, the case I do not add yet is:
- a reparentable timeline creation is interleaved with timeline_detach
    - this is worse, because the ancestor will be different on different timelines
    - deleting such timeline is fine
    - but deleting such timelines ancestor will leave the timeline deleted on some nodes and not on all
    - action here would be "detach the timeline", which will end up in a converged state

Reply:
Resolving this thread for internal slack discussion.